plugins/inputs/syslog/syslog.go

```diff
@@ -0,0 +1,378 @@
+			conn.Close()
+			continue
+		}
+		s.connections[conn.RemoteAddr().String()] = conn
+		s.connectionsMu.Unlock()
+
+		if err := s.setKeepAlive(tcpConn); err != nil {
+			acc.AddError(fmt.Errorf("unable to configure keep alive (%s): %s", s.Address, err))
+		}
+
+		go s.handle(conn, acc)
+	}
+
+	s.connectionsMu.Lock()
+	for _, c := range s.connections {
+		c.Close()
+	}
+	s.connectionsMu.Unlock()
+}
+
+func (s *Syslog) removeConnection(c net.Conn) {
+	s.connectionsMu.Lock()
+	delete(s.connections, c.RemoteAddr().String())
+	s.connectionsMu.Unlock()
+}
+
+func (s *Syslog) handle(conn net.Conn, acc telegraf.Accumulator) {
+	defer s.removeConnection(conn)
+	defer conn.Close()
+
+	if s.ReadTimeout != nil && s.ReadTimeout.Duration > 0 {
+		conn.SetReadDeadline(time.Now().Add(s.ReadTimeout.Duration))
+	}
+
+	var p *rfc5425.Parser
+	if s.BestEffort {
+		p = rfc5425.NewParser(conn, rfc5425.WithBestEffort())
+	} else {
+		p = rfc5425.NewParser(conn)
+	}
+
+	p.ParseExecuting(func(r *rfc5425.Result) {
+		s.store(*r, acc)
+	})
+}
+
+func (s *Syslog) setKeepAlive(c *net.TCPConn) error {
+	if s.KeepAlivePeriod == nil {
+		return nil
+	}
+
+	if s.KeepAlivePeriod.Duration == 0 {
+		return c.SetKeepAlive(false)
+	}
+	if err := c.SetKeepAlive(true); err != nil {
+		return err
+	}
+	return c.SetKeepAlivePeriod(s.KeepAlivePeriod.Duration)
+}
+
+func (s *Syslog) store(res rfc5425.Result, acc telegraf.Accumulator) {
+	if res.Error != nil {
+		acc.AddError(res.Error)
+	}
+	if res.MessageError != nil {
+		acc.AddError(res.MessageError)
+	}
+	if res.Message != nil {
+		acc.AddFields("syslog", fields(res.Message), tags(res.Message), s.now())
+	}
+}
+
+func tags(msg *rfc5424.SyslogMessage) map[string]string {
+	ts := map[string]string{}
+	if lvl := msg.SeverityLevel(); lvl != nil {
+		ts["severity"] = strconv.Itoa(int(*msg.Severity()))
+		ts["severity_level"] = *lvl
+	}
+
+	if f := msg.FacilityMessage(); f != nil {
+		ts["facility"] = strconv.Itoa(int(*msg.Facility()))
+		ts["facility_message"] = *f
+	}
+
+	if msg.Hostname() != nil {
+		ts["hostname"] = *msg.Hostname()
+	}
+
+	if msg.Appname() != nil {
+		ts["appname"] = *msg.Appname()
+	}
+
+	return ts
+}
+
+func fields(msg *rfc5424.SyslogMessage) map[string]interface{} {
+	flds := map[string]interface{}{
+		"version": msg.Version(),
+	}
+
+	if msg.Timestamp() != nil {
+		flds["timestamp"] = *msg.Timestamp()
+	}
+
+	if msg.ProcID() != nil {
+		flds["procid"] = *msg.ProcID()
+	}
+
+	if msg.MsgID() != nil {
+		flds["msgid"] = *msg.MsgID()
+	}
+
+	if msg.Message() != nil {
+		flds["message"] = *msg.Message()
+	}
+
+	if msg.StructuredData() != nil {
+		for sdid, sdparams := range *msg.StructuredData() {
+			if len(sdparams) == 0 {
+				// When SD-ID does not have params we indicate its presence with a bool
+				flds[sdid] = true
+				continue
+			}
+			for name, value := range sdparams {
+				// Using whitespace as separator since it is not allowed by the grammar within SDID
+				flds[sdid+" "+name] = value
```

Comment:
I've chosen to use the whitespace as a separator since this character cannot be present by specification (and so by go-syslog library) within a `SDID` or a within a `SDPARAMNAME`.
Look [here](https://github.com/influxdata/go-syslog/blob/develop/rfc5424/rfc5424.rl#L59).

If we choose to use `_` we'll not longer be able to distinguish a `SDID` without parameters, from a `SDPARAM` looking only at the field name.

In such setting, a field `abc_xzy` could represent at the same time a `SDID` without parameters -  `.. [abc_xyz] ..` - or a `SDID` with a parameter with name `xyz` - `.. [abc xyz="value"] ..`.
Thus, a user, to distinguish a posteriori, should look at the kind of the field's value - ie.,`bool` vs `string`.

Let's talk about it together pls :)

Reply:
Talking about it with @goller he suggested to create an option - eg., `sd_separator` - that defaults to `_`.

Reply:
Done.